src/cbmod/callbacks.cc

```diff
@@ -153,34 +153,29 @@ int neb::callback_acknowledgement(int callback_type, void* data) {
   try {
     // In/Out variables.
     nebstruct_acknowledgement_data const* ack_data;
-    std::shared_ptr<neb::acknowledgement> ack(new neb::acknowledgement);
+    std::shared_ptr<neb::acknowledgement> ack = std::make_shared<neb::acknowledgement>(
+        ack_data->acknowledgement_type,
+        ack_data->author_name ? ack_data->author_name : "",
+        ack_data->comment_data ? ack_data->comment_data : "",
+        time(nullptr),
+        ack_data->host_id,
+        ack_data->service_id,
+        config::applier::state::instance().poller_id(),
+        ack_data->is_sticky,
+        ack_data->notify_contacts,
+        ack_data->persistent_comment,
+        ack_data->state);
 
-    // Fill output var.
-    ack_data = static_cast<nebstruct_acknowledgement_data*>(data);
-    ack->acknowledgement_type = ack_data->acknowledgement_type;
-    if (ack_data->author_name)
-      ack->author = ack_data->author_name;
-    if (ack_data->comment_data)
-      ack->comment = ack_data->comment_data;
-    ack->entry_time = time(nullptr);
     if (!ack_data->host_id)
       throw msg_fmt("unnamed host");
     if (ack_data->service_id) {
-      ack->host_id = ack_data->host_id;
-      ack->service_id = ack_data->service_id;
       if (!ack->host_id || !ack->service_id)
```

Comment:
No use to keep this double if.
Make just one:
if (ack_data->service_id && !ack->host_id)